File: frontend/app/utils/rental/fx.ts
/** Default Cambodia market rate: how many KHR equal 1 USD. */
export const DEFAULT_USD_KHR_RATE = 4100

export type PaymentCurrency = 'USD' | 'KHR'

export function normalizePaymentCurrency(value: unknown): PaymentCurrency {
  return String(value || '').toUpperCase() === 'KHR' ? 'KHR' : 'USD'
}

export function normalizeExchangeRate(value: unknown, fallback = DEFAULT_USD_KHR_RATE): number {
  const rate = Number(value)
  if (!Number.isFinite(rate) || rate <= 0) return fallback
  return Number(rate.toFixed(4))
}

/** A cross-currency USD/KHR rate of 1 is invalid; recover with the market fallback. */
export function exchangeRateForCurrencies(
  value: unknown,
  paymentCurrency: unknown,
  rentalCurrency: unknown,
  fallback = DEFAULT_USD_KHR_RATE,
): number {
  const rate = normalizeExchangeRate(value, fallback)
  return needsExchangeRate(
    normalizePaymentCurrency(paymentCurrency),
    normalizePaymentCurrency(rentalCurrency),
  ) && rate <= 1
    ? normalizeExchangeRate(fallback)
    : rate
}

/** Convert a tendered payment into the rental's accounting currency. */
export function toRentalCurrencyAmount(
  tenderedAmount: number,
  paymentCurrency: PaymentCurrency,
  rentalCurrency: string,
  exchangeRate: number = DEFAULT_USD_KHR_RATE,
): number {
  const tendered = Math.max(0, Number(tenderedAmount) || 0)
  const rental = normalizePaymentCurrency(rentalCurrency)
  const payment = normalizePaymentCurrency(paymentCurrency)
  const rate = exchangeRateForCurrencies(exchangeRate, payment, rental)
  if (payment === rental) return Number(tendered.toFixed(2))
  if (rental === 'USD' && payment === 'KHR') return Number((tendered / rate).toFixed(2))
  if (rental === 'KHR' && payment === 'USD') return Number((tendered * rate).toFixed(2))
  return Number(tendered.toFixed(2))
}

/** Convert a rental-currency amount into the selected payment currency. */
export function fromRentalCurrencyAmount(
  rentalAmount: number,
  paymentCurrency: PaymentCurrency,
  rentalCurrency: string,
  exchangeRate: number = DEFAULT_USD_KHR_RATE,
): number {
  const amount = Math.max(0, Number(rentalAmount) || 0)
  const rental = normalizePaymentCurrency(rentalCurrency)
  const payment = normalizePaymentCurrency(paymentCurrency)
  const rate = exchangeRateForCurrencies(exchangeRate, payment, rental)
  if (payment === rental) return Number(amount.toFixed(2))
  if (rental === 'USD' && payment === 'KHR') return Number((amount * rate).toFixed(0))
  if (rental === 'KHR' && payment === 'USD') return Number((amount / rate).toFixed(2))
  return Number(amount.toFixed(2))
}

export function needsExchangeRate(paymentCurrency: PaymentCurrency, rentalCurrency: string): boolean {
  return normalizePaymentCurrency(paymentCurrency) !== normalizePaymentCurrency(rentalCurrency)
}

/**
 * Smallest editable increment for an entered amount by currency.
 * KHR has no subunit, so entered riels step by 1; USD allows cents.
 */
export function currencyInputStep(currency: unknown): number {
  return normalizePaymentCurrency(currency) === 'KHR' ? 1 : 0.01
}

/** Round an entered amount to the smallest unit of its currency. */
export function roundCurrencyAmount(amount: unknown, currency: unknown): number {
  const value = Math.max(0, Number(amount) || 0)
  return normalizePaymentCurrency(currency) === 'KHR'
    ? Math.round(value)
    : Number(value.toFixed(2))
}

/**
 * Build invoice KHR amounts from entered tendered values when available,
 * so ៛10,000 stays ៛10,000 instead of rounding back through USD as ៛10,004.
 */
export function invoiceKhrAmounts(input: {
  subtotal: number
  deposit: number
  discount?: number
  paid: number
  rentalCurrency?: string
  exchangeRate?: number
  depositTendered?: number | null
  depositCurrency?: string | null
  paidTendered?: number | null
  paidCurrency?: string | null
}): {
  subtotalKhr: number
  depositKhr: number
  discountKhr: number
  totalKhr: number
  paidKhr: number
  outstandingKhr: number
} {
  const rentalCurrency = normalizePaymentCurrency(input.rentalCurrency || 'USD')
  const rate = normalizeExchangeRate(input.exchangeRate)
  const subtotalKhr = fromRentalCurrencyAmount(input.subtotal, 'KHR', rentalCurrency, rate)

  const depositCurrency = normalizePaymentCurrency(input.depositCurrency || rentalCurrency)
  const depositTendered = Number(input.depositTendered || 0)
  const depositKhr = depositCurrency === 'KHR' && depositTendered > 0
    ? Math.round(depositTendered)
    : fromRentalCurrencyAmount(input.deposit, 'KHR', rentalCurrency, rate)
  const discountKhr = fromRentalCurrencyAmount(input.discount || 0, 'KHR', rentalCurrency, rate)

  const paidCurrency = normalizePaymentCurrency(input.paidCurrency || rentalCurrency)
  const paidTendered = Number(input.paidTendered || 0)
  const paidKhr = paidCurrency === 'KHR' && paidTendered > 0
    ? Math.round(paidTendered)
    : fromRentalCurrencyAmount(input.paid, 'KHR', rentalCurrency, rate)

  const totalKhr = Math.max(0, subtotalKhr + depositKhr - discountKhr)
  const outstandingKhr = Math.max(0, totalKhr - paidKhr)
  return { subtotalKhr, depositKhr, discountKhr, totalKhr, paidKhr, outstandingKhr }
}

File: backend/app/core/fx.py
"""USD/KHR payment conversion helpers.

``exchange_rate`` always means how many KHR equal 1 USD (Cambodia market convention).
Payment ``amount`` credited to a rental is always stored in the rental currency.
"""

from __future__ import annotations

from decimal import Decimal

from app.core.money import money

DEFAULT_USD_KHR_RATE = Decimal("4100")


def normalize_payment_currency(value: str | None) -> str:
    code = str(value or "USD").strip().upper()
    return "KHR" if code == "KHR" else "USD"


def normalize_exchange_rate(value) -> Decimal:
    try:
        rate = Decimal(str(value))
    except Exception:
        return DEFAULT_USD_KHR_RATE
    if rate <= 0:
        return DEFAULT_USD_KHR_RATE
    return rate.quantize(Decimal("0.0001"))


def exchange_rate_for_currencies(value, payment_currency: str | None, rental_currency: str | None) -> Decimal:
    """Reject the placeholder rate 1 when USD and KHR actually differ."""
    rate = normalize_exchange_rate(value)
    if normalize_payment_currency(payment_currency) != normalize_payment_currency(rental_currency) and rate <= 1:
        return DEFAULT_USD_KHR_RATE
    return rate


def to_rental_currency_amount(
    tendered_amount,
    payment_currency: str | None,
    rental_currency: str | None,
    exchange_rate=DEFAULT_USD_KHR_RATE,
) -> Decimal:
    tendered = money(tendered_amount)
    rental = normalize_payment_currency(rental_currency)
    payment = normalize_payment_currency(payment_currency)
    rate = exchange_rate_for_currencies(exchange_rate, payment, rental)
    if payment == rental:
        return tendered
    if rental == "USD" and payment == "KHR":
        return money(tendered / rate)
    if rental == "KHR" and payment == "USD":
        return money(tendered * rate)
    return tendered


def resolve_payment_money(
    *,
    rental_currency: str | None,
    payment_currency: str | None = None,
    amount=None,
    tendered_amount=None,
    exchange_rate=None,
) -> tuple[Decimal, Decimal, Decimal, str]:
    """Return ``(amount_in_rental_currency, tendered_amount, exchange_rate, payment_currency)``."""
    rental = normalize_payment_currency(rental_currency)
    payment = normalize_payment_currency(payment_currency or rental)
    rate = exchange_rate_for_currencies(
        exchange_rate if exchange_rate is not None else DEFAULT_USD_KHR_RATE,
        payment,
        rental,
    )
    if payment == rental:
        rate = Decimal("1.0000") if payment == "USD" else rate

    if tendered_amount is not None:
        tendered = money(tendered_amount)
        credited = to_rental_currency_amount(tendered, payment, rental, rate)
    elif amount is not None:
        credited = money(amount)
        if payment == rental:
            tendered = credited
        elif rental == "USD" and payment == "KHR":
            tendered = money(credited * rate)
        elif rental == "KHR" and payment == "USD":
            tendered = money(credited / rate)
        else:
            tendered = credited
    else:
        credited = Decimal("0.00")
        tendered = Decimal("0.00")

    if payment == rental and payment == "USD":
        rate = Decimal("1.0000")
    return credited, tendered, rate, payment
